encoding/bytesutil/bytes.go

```diff
@@ -383,3 +383,14 @@ func IsHex(b []byte) bool {
 	}
 	return hexRegex.Match(b)
 }
+
+// ReverseByteOrder Switch the endianness of a byte slice by reversing its order.
+// this function does not modify the actual input bytes.
+func ReverseByteOrder(input []byte) []byte {
```

Comment:
This is slow @terencechain. Consider using bitwise operations instead, as there are many popular ways of reversing endianness. See https://dev.to/brianmcgraw/reversing-bits-in-golang-1396

Reply:
the bitwise operation doesn't work too well here, given we assume the input length is arbitrary size, and I experiment capping it to length32 with the following

```
	x := [32]byte{}

	// Load in reverse order
	a := binary.BigEndian.Uint64(input[0:])
	b := binary.BigEndian.Uint64(input[8:])
	c := binary.BigEndian.Uint64(input[16:])
	d := binary.BigEndian.Uint64(input[24:])

	// Store in normal order
	binary.LittleEndian.PutUint64(x[0:], d)
	binary.LittleEndian.PutUint64(x[8:], c)
	binary.LittleEndian.PutUint64(x[16:], b)
	binary.LittleEndian.PutUint64(x[24:], a)
```

Unfortunately, I don't see much performance gain:
Original:
BenchmarkReverseByteOrder32-12        116538759            10.13 ns/op
New:
BenchmarkReverseByteOrder32-12        107376026            11.04 ns/op

What's the best way to move forward? @rauljordan and cc'ing original author @jmozah

Reply:
ah no problem, i see this is arbitrary length so we cannot make assumptions about when to stop with the bitwise operations. Thanks for the benchmark! I am ok with the for loop in that case as the lists won't be too large. I'm mostly concerned about how often this happens and how much copying needs to be done. Is there a way we can do this without copies? e.g. in-place. Maybe we can have a method that does it in-place and another which copies. If it isn't something that will be done thousands of times per second, we should be ok

Reply:
> ah no problem, i see this is arbitrary length so we cannot make assumptions about when to stop with the bitwise operations. Thanks for the benchmark! I am ok with the for loop in that case as the lists won't be too large. I'm mostly concerned about how often this happens and how much copying needs to be done. Is there a way we can do this without copies? e.g. in-place. Maybe we can have a method that does it in-place and another which copies. If it isn't something that will be done thousands of times per second, we should be ok

We can't do it in place. `ReverseByteOrder` is used to revert execution's payload's `base_fee_per_gas`. Execution payload's `base_fee_per_gas` is sent from EL, which uses big-endian to CL which uses small endian hence the conversation is required, and mutation in place does not work here. Luckily this is only called once per Bellatrix block (12s), so it's not frequent enough to warrant any concern

Reply:
> Recommended to use bitwise operators instead of a for loop

@rauljordan I agree that bitwise looks performant. Given the genericness and the frequency at which these are called, I would any day prefer something simple to understand. Something like this https://github.com/golang/go/wiki/SliceTricks#reversing. Anyway, @terencechain experiment above shows not much diff in performance. This may be because compilers heavily optimize for loops with vectorized instructions based on the processor.